modepy/shapes.py

```diff
@@ -279,51 +287,52 @@ def get_basis_with_mode_ids(shape: Shape, order: int):
 class Simplex(Shape):
     @property
     def nfaces(self):
-        return self.dims + 1
+        return self.dim + 1
 
     @property
     def nvertices(self):
         return self.dim + 1
 
 
-@get_unit_vertices.register(Simplex)
-def _(shape: Simplex):
-    from modepy.tools import unit_vertices
-    return unit_vertices(shape.dims)
+@dataclass(frozen=True)
+class _SimplexFace(Simplex, Face):
+    pass
 
 
-@get_face_vertex_indices.register(Simplex)
+@biunit_vertices_for_shape.register(Simplex)
 def _(shape: Simplex):
-    fvi = np.empty((shape.dims + 1, shape.dims), dtype=np.int)
-    indices = np.arange(shape.dims + 1)
-
-    for iface in range(shape.nfaces):
-        fvi[iface, :] = np.hstack([indices[:iface], indices[iface + 1:]])
-
-    return fvi
+    from modepy.tools import unit_vertices
+    return unit_vertices(shape.dim).T.copy()
 
 
-@get_face_map.register(Simplex)
-def _(shape: Simplex, face_vertices: np.ndarray):
-    dims, npoints = face_vertices.shape
-    if npoints != dims:
+def _simplex_face_to_vol_map(face_vertices, p: np.ndarray):
+    dim, npoints = face_vertices.shape
+    if npoints != dim:
         raise ValueError("'face_vertices' has wrong shape")
 
     origin = face_vertices[:, 0].reshape(-1, 1)
     face_basis = face_vertices[:, 1:] - origin
 
-    return lambda p: origin + np.einsum("ij,jk->ik", face_basis, (1 + p) / 2)
+    return origin + np.einsum("ij,jk->ik", face_basis, (1 + p) / 2)
 
 
-@get_quadrature.register(Simplex)
-def _(shape: Simplex, order: int):
-    import modepy as mp
-    try:
-        quad = mp.XiaoGimbutasSimplexQuadrature(2*order + 1, shape.dims)
-    except (mp.QuadratureRuleUnavailable, ValueError):
-        quad = mp.GrundmannMoellerSimplexQuadrature(order, shape.dims)
+@faces_for_shape.register(Simplex)
+def _(shape: Simplex):
+    face_vertex_indices = np.empty((shape.dim + 1, shape.dim), dtype=np.int)
+    indices = np.arange(shape.dim + 1)
 
-    return quad
+    for iface in range(shape.nfaces):
+        face_vertex_indices[iface, :] = \
+                np.hstack([indices[:iface], indices[iface + 1:]])
```

Comment:
I sort of just copy pasted this from a test case. Does it match the hardcoded `face_vertex_indices` in meshmode?

Reply:
Those are going away. My plan is to make this one big apocalypse where meshmode gives up on knowing anything that modepy can tell it.